pkg/cgu/cgu.go

```diff
@@ -338,79 +336,55 @@ func (builder *CguBuilder) WaitUntilDeleted(timeout time.Duration) error {
-	resourceCRD := "cgu"
-
-	if builder == nil {
-		glog.V(100).Infof("The %s builder is uninitialized", resourceCRD)
-
-		return false, fmt.Errorf("error: received nil %s builder", resourceCRD)
-	}
-
-	if builder.Definition == nil {
-		glog.V(100).Infof("The %s is undefined", resourceCRD)
-
-		return false, fmt.Errorf(msg.UndefinedCrdObjectErrString(resourceCRD))
-	}
-
-	if builder.apiClient == nil {
-		glog.V(100).Infof("The %s builder apiclient is nil", resourceCRD)
-
-		return false, fmt.Errorf("%s builder cannot have nil apiClient", resourceCRD)
-	}
-
-	if builder.errorMsg != "" {
-		glog.V(100).Infof("The %s builder has error message: %s", resourceCRD, builder.errorMsg)
-
-		return false, fmt.Errorf(builder.errorMsg)
-	}
-
-	return true, nil
-}
-
-// WaitUntilComplete waits the specified timeout for the CGU to complete.
-func (builder *CguBuilder) WaitUntilComplete(timeout time.Duration) (*CguBuilder, error) {
+// WaitForCondition waits until the CGU has a condition that matches the expected, checking only the Type, Status,
+// Reason, and Message fields. For the message field, it matches if the message contains the expected. Zero fields in
+// the expected condition are ignored.
+func (builder *CguBuilder) WaitForCondition(expected metav1.Condition, timeout time.Duration) (*CguBuilder, error) {
 	if valid, err := builder.validate(); !valid {
-		return builder, err
+		return nil, err
 	}
 
-	glog.V(100).Infof("Waiting for CGU %s to complete", builder.Definition.Name)
-
 	if !builder.Exists() {
 		glog.V(100).Infof("The CGU does not exist on the cluster")
 
-		return builder, fmt.Errorf(builder.errorMsg)
+		return nil, fmt.Errorf(
+			"cgu object %s does not exist in namespace %s", builder.Definition.Name, builder.Definition.Namespace)
 	}
 
-	// Polls periodically to determine if CGU is in desired state.
-	var err error
-	err = wait.PollUntilContextTimeout(
-		context.TODO(), time.Second*3, timeout, true, func(ctx context.Context) (bool, error) {
-			builder.Object, err = builder.apiClient.RanV1alpha1().ClusterGroupUpgrades(builder.Definition.Namespace).Get(
-				context.TODO(), builder.Definition.Name, metav1.GetOptions{})
+	err := wait.PollUntilContextTimeout(
+		context.TODO(), 3*time.Second, timeout, true, func(ctx context.Context) (bool, error) {
+			var err error
+			builder.Object, err = builder.apiClient.RanV1alpha1().ClusterGroupUpgrades(builder.Definition.Namespace).
+				Get(context.TODO(), builder.Definition.Name, metav1.GetOptions{})
 
 			if err != nil {
-				return false, nil
+				glog.V(100).Info("failed to get cgu %s/%s: %w", builder.Definition.Name, builder.Definition.Namespace, err)
+
+				return false, err
 			}
 
 			builder.Definition = builder.Object
 
 			for _, condition := range builder.Object.Status.Conditions {
-				if condition.Status == isTrue && condition.Type == isComplete {
-					return true, nil
+				if (expected.Type != "" && condition.Type != expected.Type) ||
```

Comment:
Do we have any particular reason to implement such complicated if condition? Why do we need to change it ?

Reply:
@kononovn The diff is a bit weird here. I'm adding a new more general method `WaitForCondition` and then reimplementing the `WaitUntilComplete` function using `WaitForCondition` so the if statement handles more now. I'll use `WaitForCondition` directly in eco-gotests as well.

I've updated the PR to split the if statement into separate ones for clarity though.